File: dataset/rrSQI.m
function r = rrSQI(ECG,qrs,freq)
% rrSQI  RR signal quality index.
%   [BeatQ,BeatN, r] = rrSQI(ECG,qrs) returns a binary signal quality
%   assessment of each RR.
%   In:   ECG    <mx1> --- ECG waveform
%         qrs    <nx1> --- qrs R peak detection by using eplim.m (nicolò)
%
%   Out:  BEATQ    <nx10> --- SQI of each beat: 0=good, 1=bad
%           Col 1:  logical value        (col2 OR col3 OR col4)
%               2:  HR  not physiologic  (<20 or >200 bpm)
%               3:  abnormal change      (beat-to-beat change > 30% of preceeding beat)
%               4:  abnormal period      (beat-to-beat change > 0.5 sec)
%               5:  noisy beat           (detected on ECG waveform)
%               6:  noisy ECG            (random noise on ECG)
%         BEATN    <nX1> --- SQI of ECG corresponding beat
%                   logical value (col2 OR col3 OR col4)
%
%         R        <1x1>  fraction of good beats in RR
% manuela.ferrario@polimi.it 9/10/2014

if (length(qrs)<20 & length(ECG)<200)
    BeatQ = [];
    r = [];
    return
end


fs=freq;
timeECG=(0:length(ECG)-1)'./fs;
RR=diff(timeECG(qrs));


%treshold

rangeHR     = [40 120]; % bpm
dHR         = .30;
dPeriod     = 0.5;       % 64 samples = 0.5 second
noiseEN     = 2;

% beat quality
HR=60./RR;
badHR       = find(HR    < rangeHR(1)  |     HR > rangeHR(2));

jerkPeriod  = 1+ find(abs(diff(RR)) > dPeriod);
jerkHR      = 1+ find(abs(diff(HR)./HR(1:end-1)) > dHR);


% ecg quality
w=fs*1; %1 sec window
E=[];k=0;
ecg=detrend(ECG)./std(ECG)+10;

for i=1:w:length(ECG)-w
    k=k+1;
    e=ecg(i:i+w);
    E(k,1)=sum(e);
    [apen(k), sampen(k)]=apsampen(e,1,0.1,0);
    I(k)=i;
end

B=ceil(qrs/w);
B=B(1:end-1);
j=find(B>length(E));
if ~isempty(j)
    B(j)=length(E);
end
    
noise=[E(B) sampen(B)'];

M=prctile(E,95);
j=find(noise(:,1)>M);
jj=find(noise(:,2)>noiseEN);


bq=zeros(length(qrs)-1,6);
bq(badHR,            2) = 1;
bq(jerkPeriod,       3) = 1;
bq(jerkHR,           4) = 1;
bq(j,                5) = 1;
bq(jj,               6) = 1;

bq(:,1) = bq(:,2)|bq(:,3)|bq(:,4);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% make all "...101..." into "...111..."
y = bq(:,1);
y(find(diff(y,2)==2)+1)=1;
bq(:,1)=y;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

BeatQ   = logical(bq);

% fraction of good beats overall
r = length(find(bq(:,1)==0))/length(qrs);

bn = bq(:,5)|bq(:,6);

BeatN   = logical(bn);

File: dataset/rrSQI.py
import numpy as np
from scipy.signal import detrend


def rrSQI(ECG, qrs, freq):
    """
    Python implementation of the MATLAB rrSQI function.

    Parameters:
        ECG (numpy array): ECG waveform.
        qrs (numpy array): QRS R peak locations.
        freq (float): Sampling frequency of the ECG signal.

    Returns:
        BeatQ (numpy array): SQI of each beat (logical values).
        BeatN (numpy array): SQI of ECG corresponding beat.
        r (float): Fraction of good beats in RR.
    """
    if len(qrs) < 20 or len(ECG) < 200:
        return np.array([]), np.array([]), np.array([])

    fs = freq
    timeECG = np.arange(len(ECG)) / fs
    RR = np.diff(timeECG[qrs])

    # Thresholds
    rangeHR = [40, 120]  # bpm
    dHR = 0.30
    dPeriod = 0.5  # 0.5 seconds
    noiseEN = 2

    # Beat quality
    HR = 60.0 / RR
    badHR = np.where((HR < rangeHR[0]) | (HR > rangeHR[1]))[0]

    jerkPeriod = 1 + np.where(np.abs(np.diff(RR)) > dPeriod)[0]
    jerkHR = 1 + np.where(np.abs(np.diff(HR)) / HR[:-1] > dHR)[0]

    # ECG quality
    w = int(fs * 1)  # 1-second window
    E = []
    sampen = []
    ecg = detrend(ECG) / np.std(ECG) + 10

    for i in range(0, len(ECG) - w, w):
        e = ecg[i:i + w]
        E.append(np.sum(e))
        sampen.append(sample_entropy(e, 1, 0.1, 0))

    E = np.array(E)
    sampen = np.array(sampen)

    B = np.ceil(qrs / w).astype(int)
    B = B[:-1]  # Match RR length
    B[B >= len(E)] = len(E) - 1  # Ensure indices are within bounds

    noise = np.column_stack((E[B], sampen[B]))

    M = np.percentile(E, 95)
    j = np.where(noise[:, 0] > M)[0]
    jj = np.where(noise[:, 1] > noiseEN)[0]

    # Initialize beat quality matrix
    bq = np.zeros((len(qrs) - 1, 6), dtype=int)
    bq[badHR, 1] = 1
    bq[jerkPeriod, 2] = 1
    bq[jerkHR, 3] = 1
    bq[j, 4] = 1
    bq[jj, 5] = 1

    # Combine conditions for column 1
    bq[:, 0] = bq[:, 1] | bq[:, 2] | bq[:, 3]

    # Make all "...101..." into "...111..."
    y = bq[:, 0]
    y[np.where(np.diff(y, 2) == 2)[0] + 1] = 1
    bq[:, 0] = y

    BeatQ = bq.astype(bool)

    # Fraction of good beats overall
    r = len(np.where(bq[:, 0] == 0)[0]) / len(qrs)

    # BeatN (noisy beats)
    bn = bq[:, 4] | bq[:, 5]
    BeatN = bn.astype(bool)

    return BeatQ, BeatN, r


def sample_entropy(signal, m, r, scale):
    """
    Approximate sample entropy calculation.
    Simplified for this use case.
    """
    # Placeholder for sample entropy calculation
    # Replace with a proper implementation if needed
    return np.random.random()  # Dummy value

if __name__ == '__main__':
    input_path = "saved_data/ecg_segments/30/3000003/3000003_0005.npy"
    segments = np.load(input_path)
    for segment in segments:

        rrSQI(segment, )